manual/manual/refman/patterns.etex

```diff
@@ -176,3 +176,25 @@ The pattern
 where \nth{c}{1}, \nth{c}{2}, \ldots, \nth{c}{n} are the characters
 that occur between \var{c} and \var{d} in the ASCII character set. For
 instance, the pattern "'0'"@'..'@"'9'" matches all characters that are digits.
+
+\subsubsection*{Exception patterns} \label{s:exception-match}
+(Introduced in OCaml 4.02)
+
+A new form of exception patterns, @ 'exception' pattern @ is allowed,
+only as a toplevel pattern under a "match"..."with" pattern-matching
+(other occurrences are rejected by the type-checker).
+
+Cases with such a toplevel pattern are called ``exception cases'',
+as opposed to regular ``value cases''.  Exception cases are applied
+when the evaluation of the matched expression raises an exception.
+The exception value is then matched against all the exception cases
+and re-raised if none of them accept the exception (as for a
+"try"..."with" block).  Since the bodies of all exception and value
+cases is outside the scope of the exception handler, they are all
+considered to be in tail-position: if the "match"..."with" block
+itself is in tail position in the current function, any function call
+in tail position in one of the case bodies results in an actual tail
+call.
+
+It is an error if all cases are exception cases in a given pattern
+matching.
```

Comment:
Might be good to have a newline at the end of the file?

Reply:
Also, I don't really understand this sentence now that I am re-reading it.

Reply:
How about:

```
It is an error if all cases in a given pattern match are exceptions. For example:

let f x = match x with
 | exception E_1 -> ...
 | exception E_2 -> ...
 | exception E_3 -> ...

is illegal.
```

A single sentence explaining the rationale for why it is illegal might also help, both by giving context and making it more obvious which case is being discussed. That might be better than having the example, which feels a bit redundant.

Reply:
Or maybe we could turn this around and say that we expect at least one value pattern. (This wording would remain correct if we add `effect p` patterns for effect handlers.)

Reply:
Perhaps:

> Beware that a pattern match must contain at least one value case.

Reply:
Or:

> A pattern match must contain at least one value case. It is an error if all cases are exceptions, because then there would be no code to handle the return of an actual value.

(Is my understanding on that correct?)

Reply:
It is probably better to be explicit here, indeed.